### PR TITLE
Dirichlet calibration

### PR DESCRIPTION
**Pull request by parmidaatg**
## Summary:
Integration of dirichlet calibration 
### Features:
- calibrated active learning 
## Checklist:

* [ x] Your code is documented (To validate this, add your module to `tests/documentation_test.py`).
* [ x] Your code is tested with unit tests.
* [x ] You moved your Issue to the PR state.

### CODE REVIEW

File: src/baal/calibration/calibration.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        self.wrapper = wrapper
+        self.init_model = deepcopy(wrapper.model)
+        self.num_classes = num_classes
+        self.loss = nn.CrossEntropyLoss()
+        self.lr = lr
+        self.reg_factor = reg_factor
+        self.mu = mu or reg_factor
+
+        # TODO: need to support kwargs for initializer for ece_per_class
+        self.wrapper.add_metric("ece", lambda: ECE())
```
Comment (Dref360) [inline src/baal/calibration/calibration.py]: Not sure I get this TODO? Can you explain :)

File: src/baal/calibration/calibration.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+        # 1. filter out unnecessary keys
+        trained_dict = {k: v for k, v in self.wrapper.state_dict().items() if k in model_dict}
+        # 2. overwrite entries in the existing state dict
+        model_dict.update(trained_dict)
+        # 3. load the new state dict
+        self.init_model.load_state_dict(model_dict)
+
+        # make sure that we are not retraining the model
+        for param in self.init_model.parameters():
```
Comment (Dref360) [inline src/baal/calibration/calibration.py]: Can we just create an optimizer with only the linear layer as parameters?

File: src/baal/modelwrapper.py [hunk 3]
```diff
@@ -256,16 +226,38 @@ def predict_on_dataset(self, dataset: Dataset, batch_size: int, iterations: int,
...
+            use_cuda (bool): If "True" will train on GPU.
+            double_fit (bool): If "True" would fit twice on the train set.
+
+        Returns:
+            loss_history (list[float]): List of loss values for each epoch.
+            weights (dict): Model weights.
+
+        """
+
+        loss_history, weights = self.calibrator.calibrate(train_loader, val_loader,
```
Comment (Dref360) [inline src/baal/modelwrapper.py]: calibrator can be None

File: src/baal/calibration/calibration.py (hunk 1 above)
Comment (parmidaatg) [reply src/baal/calibration/calibration.py]: `ece_per_cls` gets `num_classes` for its initializer. Currently the `add_metric` initializer doesn't support input variables

Comment (Dref360) [inline src/baal/calibration/calibration.py]:
```suggestion
        self.wrapper.add_metric("ece", lambda: ECE(num_classes))
```

This should work @parmidaatg ?

File: experiments/vgg_mcdropout_cifar10.py [hunk 4]
```diff
@@ -102,7 +102,6 @@ def main():
                                      model.predict_on_dataset,
                                      heuristic,
                                      hyperparams.get('n_data_to_label', 1),
-                                     batch_size=10,
```
Comment (Dref360) [inline experiments/vgg_mcdropout_cifar10.py]: Revert

File: src/baal/calibration/calibration.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        Returns:
+            The regularization term for the linear layer.
+        """
+        name, params = list(self.model.named_parameters())[-2]
+        assert 'weight' in name
+        w_l2_factor = params.norm(2)
+
+        name, params = list(self.model.named_parameters())[-1]
+        assert 'bias' in name
+        b_l2_factor = params.norm(2)
```
Comment (Dref360) [inline src/baal/calibration/calibration.py]:
```suggestion
        weight_p, bias_p = self.dirichlet_linear.parameters()
        w_l2_factor = weight_p.norm(2)
        b_l2_factor = bias_p.norm(2)
``

File: src/baal/calibration/calibration.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        optimizer = Adam(self.dirichlet_linear.parameters(), lr=self.lr)
+        loss_history, weights = self.wrapper.train_and_test_on_datasets(train_set, test_set,
+                                                                        optimizer, batch_size,
+                                                                        epoch, use_cuda,
+                                                                        return_best_weights=True,
+                                                                        patience=None, **kwargs)
+        self.init_model.load_state_dict(weights)
+
+        if double_fit:
+            self.lr = self.lr / 10
```
Comment (Dref360) [inline src/baal/calibration/calibration.py]: This will cause the learning rate to approach 0 if called multiple time.

File: src/baal/modelwrapper.py [hunk 7]
```diff
@@ -26,14 +26,17 @@ class ModelWrapper:
         model (nn.Module): The model to optimize.
         criterion (Callable): A loss function.
         replicate_in_memory (bool): Replicate in memory optional.
+        calibrator (object): Predefined class.
```
Comment (Dref360) [inline src/baal/modelwrapper.py]: Should we include that in ModelWrapper?

File: src/baal/calibration/calibration.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    Source: https://arxiv.org/abs/1910.12656
+
+    Args:
+
+        wrapper (ModelWrapper): Provides training and testing methods.
+        num_classes (int): Number of classes in classification task.
+        lr (float): Learning rate.
+        reg_factor (float): Regularization factor for the linear layer weights.
+        mu (float): Regularization factor for the linear layer biases.
+            If not given, will be initialized by "l".
```
Comment (Dref360) [inline src/baal/calibration/calibration.py]: "l" == reg_factor?

File: src/baal/modelwrapper.py (hunk 7 above)
Comment (parmidaatg) [reply src/baal/modelwrapper.py]: could you explain?

File: src/baal/calibration/calibration.py (hunk 8 above)
Comment (parmidaatg) [reply src/baal/calibration/calibration.py]: yeah, flake was annoyed by "l" so I changed it to "red_factor"

File: src/baal/modelwrapper.py [hunk 9]
```diff
@@ -28,7 +28,8 @@ class ModelWrapper:
         replicate_in_memory (bool): Replicate in memory optional.
     """
 
-    def __init__(self, model, criterion, replicate_in_memory=True):
+    def __init__(self, model, criterion,
```
Comment (Dref360) [inline src/baal/modelwrapper.py]: I think we can revert this file.
